packages/taiko-client/proposer/profitability.go

```diff
@@ -0,0 +1,389 @@
+package proposer
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sort"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/txmgr"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/taikoxyz/taiko-mono/packages/taiko-client/pkg/utils"
+)
+
+// CostEstimator is an interface for estimating the L1 cost of proposing a batch.
+type CostEstimator interface {
+	estimateL1Cost(ctx context.Context, candidate *txmgr.TxCandidate) (*big.Int, error)
+}
+
+// isProfitable checks if proposing the given transaction batch is profitable.
+// It performs profitability checks and can adjust the base fee and filter transactions if needed.
+// Returns (isProfitable bool, baseFeeAdjusted bool, error)
+func (p *Proposer) isProfitable(
+	ctx context.Context,
+	txBatch *[]types.Transactions,
+	l2BaseFee **big.Int,
+	candidate *txmgr.TxCandidate,
+	txs *uint64,
+) (bool, bool, error) {
+	estimatedCost, err := p.costEstimator.estimateL1Cost(ctx, candidate)
+	if err != nil {
+		return false, false, fmt.Errorf("failed to estimate L1 cost: %w", err)
+	}
+
+	// Compute collected fees for the original batch/base fee
+	originalCollectedFees := p.computeL2Fees(*txBatch, *l2BaseFee)
+
+	log.Info("Profitability check (standard base fee)",
+		"estimatedCost", utils.WeiToEther(estimatedCost),
+		"collectedFees", utils.WeiToEther(originalCollectedFees),
+		"l2BaseFee", utils.WeiToEther(*l2BaseFee),
+		"numBatches", len(*txBatch),
+		"numTransactions", *txs,
+	)
+
+	// We'll keep track of the best option found (highest collected fees)
+	bestCollectedFees := new(big.Int).Set(originalCollectedFees)
+	bestTxBatch := *txBatch
+	bestL2BaseFee := new(big.Int).Set(*l2BaseFee)
+	bestTxs := *txs
+	bestAdjusted := false
+
+	// Use optimized algorithm to find the best base fee threshold
+	// This is O(n log n) instead of O(n²)
+	optimalBaseFee, optimalCollectedFees := p.findOptimalBaseFeeThreshold(*txBatch)
+
+	if optimalBaseFee != nil && optimalCollectedFees.Cmp(bestCollectedFees) > 0 {
+		// Filter transactions with the optimal base fee
+		filteredTxBatch := p.filterTxsByBaseFee(*txBatch, optimalBaseFee)
+		filteredTxCount := countTxsInBatch(filteredTxBatch)
+
+		// Calculate improvement percentage
+		improvementPct := float64(optimalCollectedFees.Int64()-originalCollectedFees.Int64()) * 100 /
+			float64(originalCollectedFees.Int64())
```

Comment:
Assuming truncation here are accepted as it's only for logging and an int64 overflow would only happen at around 9 ETH.